[CF-1266]: geojson polygon filtering

bjungs opened the pull request:
In CityFlows we have a requirement to be able to filter the datasets based on what the user selects on the screen. Currently, KeplerGL supports filtering datasets using Polygons drawn to the screen. However, GeoJson data types are not currently supported.

The goal of this PR is to add support to filtering GeoJson data using Polygons (at least to the feature types we need at the moment).

[CityFlows-geojson-filtering.webm](https://user-images.githubusercontent.com/16376552/201674366-d774533f-4b8e-43bd-944e-4d6bb66a1d22.webm)

bjungs committed:
Adding support for Polygon Filtering of GeoJson features

bjungs committed:
Refactoring for better clarity

bjungs committed:
Added tests for isGeoJsonPositionInPolygon

- testing Point positions

bjungs committed:
Added tests for isGeoJsonPositionInPolygon

- testing LineString positions

bjungs committed:
Added tests for isGeoJsonPositionInPolygon

- testing Polygon positions

ismakutl reviewed (changes requested):
Hey! first of all I want to mention it's a very nice addition!

I noticed you ran yarn and re-generated the lock file, I would revert that change because nothing changed in the package.json I see, otherwise our pipeline might break

Would be nice to see this in action, by providing an example in `examples/demo/app.js`

braemJef commented:
@ismakutl 
> Hey! first of all I want to mention it's a very nice addition!
> 
> I noticed you ran yarn and re-generated the lock file, I would revert that change because nothing changed in the package.json I see, otherwise our pipeline might break
> 
> Would be nice to see this in action, by providing an example in `examples/demo/app.js`

I think this is a mistake on our end, as we should always commit and push this file. The fact that it updated for him when doing a yarn install meant that we did not do a good job at this.

I propose we do in fact keep the file in the PR.

ismakutl commented:
> I think this is a mistake on our end, as we should always commit and push this file. The fact that it updated for him when doing a yarn install meant that we did not do a good job at this.
> 
> I propose we do in fact keep the file in the PR.

We never installed any external libraries until now, at least not that I'm aware of
These libraries are installed with Node v12, if @bjungs has Node v14-18, it's pretty normal the lock is re-generated, so I would avoid making such changes

bjungs commented:
> Hey! first of all I want to mention it's a very nice addition!
> 
> I noticed you ran yarn and re-generated the lock file, I would revert that change because nothing changed in the package.json I see, otherwise our pipeline might break
> 
> Would be nice to see this in action, by providing an example in `examples/demo/app.js`

Thanks for your feedback. I'll look into the demo but I'm pretty sure this can already be demoed with the sample data.



> > I think this is a mistake on our end, as we should always commit and push this file. The fact that it updated for him when doing a yarn install meant that we did not do a good job at this.
> > I propose we do in fact keep the file in the PR.
> 
> We never installed any external libraries until now, at least not that I'm aware of These libraries are installed with Node v12, if @bjungs has Node v14-18, it's pretty normal the lock is re-generated, so I would avoid making such changes

Perhaps we can use `engine-strict` configuration in `.npmrc` file  to enforce the specific versions used here. I used Node 18 indeed. Which version should I use instead? 12?

ismakutl commented:
> > Hey! first of all I want to mention it's a very nice addition!
> > I noticed you ran yarn and re-generated the lock file, I would revert that change because nothing changed in the package.json I see, otherwise our pipeline might break
> > Would be nice to see this in action, by providing an example in `examples/demo/app.js`
> 
> Thanks for your feedback. I'll look into the demo but I'm pretty sure this can already be demoed with the sample data.
> 
> > > I think this is a mistake on our end, as we should always commit and push this file. The fact that it updated for him when doing a yarn install meant that we did not do a good job at this.
> > > I propose we do in fact keep the file in the PR.
> > 
> > 
> > We never installed any external libraries until now, at least not that I'm aware of These libraries are installed with Node v12, if @bjungs has Node v14-18, it's pretty normal the lock is re-generated, so I would avoid making such changes
> 
> Perhaps we can use `engine-strict` configuration in `.npmrc` file to enforce the specific versions used here. I used Node 18 indeed. Which version should I use instead? 12?

Sounds good! Indeed kepler is built with Node 12, see the package.json https://github.com/imec-int/kepler.gl/blob/v2.5.5-master/package.json, normally you shouldn't see any changes in the lock file

bjungs commented:
Oh I didn't see the engines there my bad. Anyway, I'll use Node 12 and rerun the build. Thanks!

bjungs commented:
Even with Node 12 the lock file is changed completely. It seems that the 'problem' is the Yarn version I'm using (the latest - v3). This causes some significant changes in the lock file as well as extra output (yarnrc.yml file and .yarn folder). Since I don't know much about Yarn itself and the migration is beyond the scope of this PR, I will simply remove the changes to the the lock file in this PR.

For more information:

https://yarnpkg.com/getting-started/qa#why-should-you-upgrade-to-yarn-modern
https://yarnpkg.com/getting-started/migration

bjungs committed:
Added support for MultiPolygon and Feature geojson types

bjungs committed:
Added tests for MultiPolygon as a Feature

bjungs committed:
Added tests for other geometries as a Features

bjungs commented:
Added support for `MultiPolygon` and `Feature` as well so that the filtering works with the sample data of `county_unemployment`.

(forgot to record the mouse this time, sorry :p)

[multipolygon-features.webm](https://user-images.githubusercontent.com/16376552/201887062-34c337aa-7595-4466-b3ac-3a0ce9f59819.webm)

bjungs committed:
Broke down tests into geometries

bjungs committed:
Added support for position as string

It seems that in some cases the getPositionAccessor function may return a string instead of an object. From the `movement_pittsburgh` example it seems that it happens with CSV files, but I'm not 100% sure if that's always the case.

In any case, it shouldn't hurt to try and parse the string value, and it works just fine in the scenario described above.

bjungs commented:
@ismakutl now you should be able to try it out with either the `movement_pittsburgh`, `county_unemployment`, or `world_flights` samples. They all use geojson data.

This should be enough to satisfy your request. Please let me know if you'd like something more specific.

Thanks!

ismakutl reviewed (approved):
LGTM!

bjungs commented on src/utils/filter-utils.js:
```diff
@@ -391,11 +391,65 @@ export const getPolygonFilterFunctor = (layer, filter, dataContainer) => {
+      return getGeoJsonPoints(position.geometry);
+    default:
+      Console.log(`Unsupported geojson type ${position?.type}`);
+      // Console.error(`Unsupported geojson type`);
```
Superfluous line

```suggestion
```

bjungs committed:
Removed superfluous comment

bjungs commented:
@ismakutl can I create a release for this? We'll need to deploy it very soon.
